_source/_posts/2023-03-20-spa-auth-tokens.md

```diff
@@ -0,0 +1,228 @@
+
+Use the following command in the terminal to get going quickly. There's a separate command for each Angular, React, and Vue.
+
+**Angular**
+```console
+okta start angular
+```
+
+**React**
+```console
+okta start react
+```
+
+**Vue**
+```console
+okta start vue
+```
+
+Follow the instructions in the terminal to start the application. You should be able to sign in and sign out. Additionally, you can watch network requests for the calls it makes.
+
+We don't have a refresh token by default. You can enable that in the Okta Admin Console. Navigate to the [Okta Developer site](https://developer.okta.com/) and sign in to your Okta organization. In the Admin Console, navigate to **Applications** > **Applications** and select the Okta application for the SPA you created. Edit the settings to enable **Refresh Token**. It automatically adds Refresh Token rotation. Save your changes.
+
+If you haven't already done so, open the SPA code in your favorite IDE and navigate to the `config.js` or `app.config.ts` file, depending on your framework. You'll see the OIDC config and a property for `scopes` where you will add `offline_access`. Your OIDC config looks something like this:
+
+```js
+{
+    clientId: CLIENT_ID,
+    issuer: ISSUER,
+    redirectUri: window.location.origin + '/login/callback',
+    scopes: ['openid', 'profile', 'email', 'offline_access']
+  }
+```
+
+Now try rerunning the SPA. If you inspect the network request, you'll see the refresh token too. You can see the tokens by looking at the contents of your Local Storage too.
+
+## Use auth tokens in SPAs
+
+Now that we have these tokens in our SPA, what do we do with them? We should let a trusted OIDC library, such as the Okta SDKs we're using, handle all the token requests and refresh them for us. Behind the scenes the OIDC library is hard at work exchanging tokens.
+
+{% img blog/spa-auth-tokens/refresh-token.jpg alt:"Refresh token call to authorization server and returning new tokens" width:"800" %}{: .center-image }
+
+We're primarily interested in the contents of the ID token and using the Access token.
+
+### How to use the Access Token
+The Access token gives us access to resources. That means we use Access tokens in outgoing API requests that our application needs, like for Sunny to get their next rawring lesson.
+
+{% img blog/spa-auth-tokens/access-token.jpg alt:"Sunny starting up a new lesson makes a call to the Rawr site's /lesson endpoint passing in the Access token" width:"800" %}{: .center-image }
+
+You'll send the Access token as a Bearer token in the HTTP call's Authorization header. So in the case of Sunny, the outgoing HTTP call looks like this.
+
+```http
+GET /lessons
+Authorization: Bearer base64encoded_access_token
```

Comment:
```suggestion
Authorization: Bearer access_token_value
```

I would avoid mentioning base64 encoding here since that could get confusing. Technically the JWT access token is base64 encoded, but only because it's a JWT, and that isn't something the developer needs to do. The important thing is to make sure they don't modify the token value that they get back from the token endpoint when they use it.